internal/controller/apiextensions/claim/configurator.go

```diff
@@ -63,7 +64,11 @@ func ConfigureComposite(_ context.Context, cm resource.CompositeClaim, cp resour
 	// external name (even if that external name was empty) in order to ensure
```

Comment:
I'm a bit confused about why we needed this logic at all. @negz `meta.AddAnnotations(cp, cm.GetAnnotations())` will not override the external name annotation if the claim doesn't have it. If it does, shouldn't it be respected? Let's say I have `bucket-name-test` as my external-name annotation in my claim so that I can get a bucket with that name, should I even be paired with a composite resource whose external name is different than that?

Reply:
My other comment also depends on this one. If we don't respect claim's annotation, then it's not late-init, hence keeping the functionality as is and only checking for `""` or `nil`(current change) would be sufficient.

Reply:
I think this was addressing the (exceedingly rare) edge case in which:

1. An XR is statically provisioned.
2. An XRC later explicitly claims the XR by setting its `spec.resourceRef`.
3. The XR and XRC both have different external-name annotations set at claim time.

In this case the XR's external name has probably already been applied. We can't just propagate the XRC's external name over the top of the XR's because doing so would presumably trickle down to managed resources and attempt to rename them in place, which almost certainly wouldn't work in most cases. So the idea here is that we basically in this edge case say "nope, we're ignoring the XRC's requested external name because the XR already has one".

In this specific edge case we probably do want to overwrite the XRC's external name with the real external name - the one from the XR. Alternately, I could get behind your suggestion that we should just refuse to let an XRC claim an XR if they have conflicting external names. Or going even further, generally don't allow a claim for a statically provisioned XR to ever specify an external name, since it would realistically never be honored.

Reply:
Hmm I can see two options then:
* Seems like this is the only place we account for external name annotation at this level. We could possibly not do anything special for external-name annotation on XRC-XR level, treat it just like any other annotation.
* Late initialize external-name annotation if the claim has empty external name annotation AND refuse to bind if both XRC and XR have non-empty external-names that are not same.

I can go either way but feel closer to the second option because we'd like composite to be similar to managed resources so that it's easier to treat it like one in multi-layered `Composition`s. @ulucinar what are your thoughts?

> Or going even further, generally don't allow a claim for a statically provisioned XR to ever specify an external name, since it would realistically never be honored.

Hmm I'd say that if user manually chose an XR with `resourceRef` and the external-name annotations of XRC and XR matches, it would be safe to bind the two.

Reply:
> I can go either way but feel closer to the second option.

I agree.

Reply:
I don't have a strong opinion here, but in the light of https://github.com/crossplane/crossplane/issues/2350#issuecomment-856203139
my understanding is that if an XRC explicitly claims a statically provisioned XR by referencing it (via its `spec.resourceRef`) and the external-name annotations in XRC and XR are in conflict, it seems to me that we had better honor the binding request communicated by setting `spec.resourceRef` both because of a desire to behave in a backwards-compatible manner and because this is an explicit communication of an intent although it's conflicting with what's specified in the XRC's external-name annotation. What's more, considering the common use cases of annotations, it would, in my opinion, make sense to reconcile the annotation in XRC to reflect the annotation specified in the referenced (statically provisioned) XR, i.e., let the referenced XR override external-name annotation specified in XRC.

Please note that if we choose to do so, for a dynamically provisioned XR, the external-name annotation will initially be set to what's specified in the claim, assuming one is set. Later in the runtime, if an external entity explicitly sets the external-name on the XR to something else, we would be propagating it back to XR.